runtime error: slice bounds out of range

My server crash after 10-20 min after start.
caller code:
```go
var gCompressPool5 sync.Pool
func mustFlateCompressWithBufferToBufWPool5(inb []byte,bufW *bytes.Buffer) {
	var flateW *flate.Writer
	obj:=gCompressPool5.Get()
	if obj==nil{
		var err error
		flateW,err = flate.NewWriter(bufW,4)
		if err!=nil{
			panic(err)
		}
	}else{
		flateW = obj.(*flate.Writer)
		flateW.Reset(bufW)
	}
	_, err := flateW.Write(inb)
	if err != nil {
		flateW.Close()
		flateW.Reset(nil)
		gCompressPool5.Put(flateW)
		panic(err)
	}
	err = flateW.Close()
	flateW.Reset(nil)
	gCompressPool5.Put(flateW)
	if err != nil {
		panic(err)
	}
}
```

```
runtime error: slice bounds out of range

goroutine 82 [running]:
github.com/klauspost/compress/flate.(*fastGen).matchlenLong(0xc0288e8000, 0x5386ca610003395c, 0xc028aea000, 0x3fffc, 0x100000, 0x13)
	github.com/klauspost/compress/flate/fast_encoder.go:208 +0xb6
github.com/klauspost/compress/flate.(*fastEncL4).Encode(0xc0288e8000, 0xc02bcf2090, 0xc02bf88000, 0xffff, 0xffff)
	github.com/klauspost/compress/flate/level4.go:127 +0x4db
github.com/klauspost/compress/flate.(*compressor).storeFast(0xc02bcf2000)
	github.com/klauspost/compress/flate/deflate.go:590 +0x1c1
github.com/klauspost/compress/flate.(*compressor).write(0xc02bcf2000, 0xc02176bffc, 0x1c80ae, 0x18a004, 0x113d380, 0xc007352960, 0xc007352960)
	github.com/klauspost/compress/flate/deflate.go:614 +0x81
github.com/klauspost/compress/flate.(*Writer).Write(0xc02bcf2000, 0xc02172c000, 0x1c80ae, 0x1ca000, 0xc000d2b568, 0x44b894, 0xc015512199)
	github.com/klauspost/compress/flate/deflate.go:781 +0x4b
...
```

I have tried a single thread, it crash like this too(after one hour):
```go
var gCompressWriter *flate.Writer
var gCompressWriterLocker sync.Mutex
func mustFlateCompressWithBufferToBufWPool3(inb []byte,bufW *bytes.Buffer) {
	gCompressWriterLocker.Lock()
	defer gCompressWriterLocker.Unlock()
	if gCompressWriter==nil{
		var err error
		gCompressWriter,err = flate.NewWriter(bufW,4)
		if err!=nil{
			panic(err)
		}
	}else{
		gCompressWriter.Reset(bufW)
	}
	_, err := gCompressWriter.Write(inb)
	if err != nil {
		gCompressWriter.Close()
		gCompressWriter.Reset(nil)
		panic(err)
	}
	err = gCompressWriter.Close()
	gCompressWriter.Reset(nil)
	if err != nil {
		panic(err)
	}
}
```

I have tried flate from golang 1.11.6, it never crash after 3 hours. (but use more cpu resource than this library)
I have tried flate from https://github.com/klauspost/compress and do not reuse the writer object, it never crash after 12 hours.(but use more cpu resource) 

Workaround right now: use flate from golang 1.11.6 and sync.Pool

When I set debugDecode to true, I got:
```
panic: t >=s:2147411689 14691

goroutine 60 [running]:
github.com/klauspost/compress/flate.(*fastGen).matchlenLong(0xc0356e0000, 0x7ffee6e900003963, 0xc03583c000, 0x78c4, 0x100000, 0x46)
	github.com/klauspost/compress/flate/fast_encoder.go:195 +0x596
github.com/klauspost/compress/flate.(*fastEncL4).Encode(0xc0356e0000, 0xc034f6a090, 0xc0357e2000, 0x78c4, 0xffff)
	github.com/klauspost/compress/flate/level4.go:127 +0x4db
github.com/klauspost/compress/flate.(*compressor).storeFast(0xc034f6a000)
	github.com/klauspost/compress/flate/deflate.go:590 +0x1c1
github.com/klauspost/compress/flate.(*compressor).close(0xc034f6a000, 0x0, 0x0)
	github.com/klauspost/compress/flate/deflate.go:715 +0x4e
github.com/klauspost/compress/flate.(*Writer).Close(0xc034f6a000, 0xc02aca8000, 0x78c4)
	github.com/klauspost/compress/flate/deflate.go:801 +0x2d
```

another crash stack from master branch. (not from #218 )
```
panic: runtime error: slice bounds out of range

goroutine 90 [running]:
github.com/klauspost/compress/flate.(*fastEncL4).Encode(0xc03089e000, 0xc0320c2090, 0xc02a4aa000, 0x8cc5, 0xffff)
        github.com/klauspost/compress/flate/level4.go:111 +0x9f4
github.com/klauspost/compress/flate.(*compressor).storeFast(0xc0320c2000)
        github.com/klauspost/compress/flate/deflate.go:590 +0x1c1
github.com/klauspost/compress/flate.(*compressor).close(0xc0320c2000, 0x0, 0x0)
        github.com/klauspost/compress/flate/deflate.go:715 +0x4e
github.com/klauspost/compress/flate.(*Writer).Close(0xc0320c2000, 0xc033692000, 0x8cc5)
        github.com/klauspost/compress/flate/deflate.go:801 +0x2d
```

Sorry, but it crashed from code #218 
```
panic: e.cur < 0: -2147281275

goroutine 78 [running]:
github.com/klauspost/compress/flate.(*fastEncL4).Encode(0xc05ab1e000, 0xc014570090, 0xc044928000, 0xffff, 0xffff)
	github.com/klauspost/compress/flate/level4.go:17 +0xaf0
github.com/klauspost/compress/flate.(*compressor).storeFast(0xc014570000)
	github.com/klauspost/compress/flate/deflate.go:590 +0x1c1
github.com/klauspost/compress/flate.(*compressor).write(0xc014570000, 0xc04fad3fee, 0x245f98, 0x126012, 0x113e5c0, 0xc03b9af5c0, 0xc03b9af5c0)
	github.com/klauspost/compress/flate/deflate.go:614 +0x81
github.com/klauspost/compress/flate.(*Writer).Write(0xc014570000, 0xc04f9b4000, 0x245f98, 0x246000, 0xc00057b568, 0x44b894, 0xc038d3a2e1)
	github.com/klauspost/compress/flate/deflate.go:781 +0x4b
```

--- a/flate/fast_encoder.go
+++ b/flate/fast_encoder.go
@@ -44,7 +44,7 @@ const (
 
 	bTableBits   = 18                                           // Bits used in the big tables
 	bTableSize   = 1 << bTableBits                              // Size of the table
-	allocHistory = maxMatchOffset * 10                          // Size to preallocate for history.
+	allocHistory = maxStoreBlockSize * 20                       // Size to preallocate for history.
 	bufferReset  = (1 << 31) - allocHistory - maxStoreBlockSize // Reset the buffer offset when reaching this.
 )
 
@@ -210,9 +210,8 @@ func (e *fastGen) matchlenLong(s, t int32, src []byte) int32 {
 
 // Reset the encoding table.
 func (e *fastGen) Reset() {
-	const historyAllocMin = maxMatchOffset * 8
-	if cap(e.hist) < historyAllocMin {
-		e.hist = make([]byte, 0, 1<<20)
+	if cap(e.hist) < allocHistory {
+		e.hist = make([]byte, 0, allocHistory)
 	}
 	// We offset current position so everything will be out of reach.
 	// If we are above the buffer reset it will be cleared anyway since len(hist) == 0.

--- a/flate/fast_encoder.go
+++ b/flate/fast_encoder.go
@@ -210,16 +210,15 @@ func (e *fastGen) matchlenLong(s, t int32, src []byte) int32 {
 
 // Reset the encoding table.
 func (e *fastGen) Reset() {
-	if cap(e.hist) < int(maxMatchOffset*8) {
-		l := maxMatchOffset * 8
-		// Make it at least 1MB.
-		if l < 1<<20 {
-			l = 1 << 20
-		}
-		e.hist = make([]byte, 0, l)
+	const historyAllocMin = maxMatchOffset * 8
+	if cap(e.hist) < historyAllocMin {
+		e.hist = make([]byte, 0, 1<<20)
 	}
-	// We offset current position so everything will be out of reach
-	e.cur += maxMatchOffset + int32(len(e.hist))
+	// We offset current position so everything will be out of reach.
+	// If we are above the buffer reset it will be cleared anyway since len(hist) == 0.
+	if e.cur <= bufferReset {
+		e.cur += maxMatchOffset + int32(len(e.hist))
+	}
 	e.hist = e.hist[:0]
 }
 

--- a/flate/fast_encoder.go
+++ b/flate/fast_encoder.go
@@ -42,10 +42,10 @@ const (
 	baseMatchLength = 3              // The smallest match length per the RFC section 3.2.5
 	maxMatchOffset  = 1 << 15        // The largest match offset
 
-	bTableBits   = 18                                           // Bits used in the big tables
-	bTableSize   = 1 << bTableBits                              // Size of the table
-	allocHistory = maxStoreBlockSize * 20                       // Size to preallocate for history.
-	bufferReset  = (1 << 31) - allocHistory - maxStoreBlockSize // Reset the buffer offset when reaching this.
+	bTableBits   = 18                                               // Bits used in the big tables
+	bTableSize   = 1 << bTableBits                                  // Size of the table
+	allocHistory = maxStoreBlockSize * 20                           // Size to preallocate for history.
+	bufferReset  = (1 << 31) - allocHistory - maxStoreBlockSize - 1 // Reset the buffer offset when reaching this.
 )
 
 const (

--- a/flate/level1.go
+++ b/flate/level1.go
@@ -1,5 +1,7 @@
 package flate
 
+import "fmt"
+
 // fastGen maintains the table for matches,
 // and the previous byte block for level 2.
 // This is the generic implementation.
@@ -14,6 +16,9 @@ func (e *fastEncL1) Encode(dst *tokens, src []byte) {
 		inputMargin            = 12 - 1
 		minNonLiteralBlockSize = 1 + 1 + inputMargin
 	)
+	if debugDecode && e.cur < 0 {
+		panic(fmt.Sprint("e.cur < 0: ", e.cur))
+	}
 
 	// Protect against e.cur wraparound.
 	for e.cur >= bufferReset {

--- a/flate/level2.go
+++ b/flate/level2.go
@@ -1,5 +1,7 @@
 package flate
 
+import "fmt"
+
 // fastGen maintains the table for matches,
 // and the previous byte block for level 2.
 // This is the generic implementation.
@@ -15,6 +17,10 @@ func (e *fastEncL2) Encode(dst *tokens, src []byte) {
 		inputMargin            = 12 - 1
 		minNonLiteralBlockSize = 1 + 1 + inputMargin
 	)
+
+	if debugDecode && e.cur < 0 {
+		panic(fmt.Sprint("e.cur < 0: ", e.cur))
+	}
 
 	// Protect against e.cur wraparound.
 	for e.cur >= bufferReset {

--- a/flate/level3.go
+++ b/flate/level3.go
@@ -1,5 +1,7 @@
 package flate
 
+import "fmt"
+
 // fastEncL3
 type fastEncL3 struct {
 	fastGen
@@ -12,6 +14,10 @@ func (e *fastEncL3) Encode(dst *tokens, src []byte) {
 		inputMargin            = 8 - 1
 		minNonLiteralBlockSize = 1 + 1 + inputMargin
 	)
+
+	if debugDecode && e.cur < 0 {
+		panic(fmt.Sprint("e.cur < 0: ", e.cur))
+	}
 
 	// Protect against e.cur wraparound.
 	for e.cur >= bufferReset {

--- a/flate/level4.go
+++ b/flate/level4.go
@@ -13,7 +13,9 @@ func (e *fastEncL4) Encode(dst *tokens, src []byte) {
 		inputMargin            = 12 - 1
 		minNonLiteralBlockSize = 1 + 1 + inputMargin
 	)
-
+	if debugDecode && e.cur < 0 {
+		panic(fmt.Sprint("e.cur < 0: ", e.cur))
+	}
 	// Protect against e.cur wraparound.
 	for e.cur >= bufferReset {
 		if len(e.hist) == 0 {

--- a/flate/level5.go
+++ b/flate/level5.go
@@ -13,6 +13,9 @@ func (e *fastEncL5) Encode(dst *tokens, src []byte) {
 		inputMargin            = 12 - 1
 		minNonLiteralBlockSize = 1 + 1 + inputMargin
 	)
+	if debugDecode && e.cur < 0 {
+		panic(fmt.Sprint("e.cur < 0: ", e.cur))
+	}
 
 	// Protect against e.cur wraparound.
 	for e.cur >= bufferReset {

--- a/flate/level6.go
+++ b/flate/level6.go
@@ -13,6 +13,9 @@ func (e *fastEncL6) Encode(dst *tokens, src []byte) {
 		inputMargin            = 12 - 1
 		minNonLiteralBlockSize = 1 + 1 + inputMargin
 	)
+	if debugDecode && e.cur < 0 {
+		panic(fmt.Sprint("e.cur < 0: ", e.cur))
+	}
 
 	// Protect against e.cur wraparound.
 	for e.cur >= bufferReset {

--- a/flate/writer_test.go
+++ b/flate/writer_test.go
@@ -10,6 +10,7 @@ import (
 	"fmt"
 	"io"
 	"io/ioutil"
+	"math"
 	"math/rand"
 	"runtime"
 	"strconv"
@@ -262,6 +263,72 @@ func TestWriteError(t *testing.T) {
 				return
 			}
 		}
+	}
+}
+
+// Test if errors from the underlying writer is passed upwards.
+func TestWriter_Reset(t *testing.T) {
+	buf := new(bytes.Buffer)
+	n := 65536
+	if !testing.Short() {
+		n *= 4
+	}
+	for i := 0; i < n; i++ {
+		fmt.Fprintf(buf, "asdasfasf%d%dfghfgujyut%dyutyu\n", i, i, i)
+	}
+	in := buf.Bytes()
+	for l := 0; l < 10; l++ {
+		l := l
+		t.Run(fmt.Sprintf("level-%d", l), func(t *testing.T) {
+			t.Parallel()
+			for offset := 1; offset <= 256; offset *= 2 {
+				// Fail after 'fail' writes
+				if testing.Short() {
+					// Skip...
+					offset = 128
+				}
+				w, err := NewWriter(ioutil.Discard, l)
+				if err != nil {
+					t.Fatalf("NewWriter: level %d: %v", l, err)
+				}
+				if w.d.fast == nil {
+					t.Skip("Not Fast...")
+					return
+				}
+				for i := 0; i < (bufferReset-len(in)-offset-maxMatchOffset)/maxMatchOffset; i++ {
+					// skip ahead to where we are close to wrap around...
+					w.d.fast.Reset()
+				}
+				w.d.fast.Reset()
+				_, err = w.Write(in)
+				if err != nil {
+					t.Fatal(err)
+				}
+				for i := 0; i < 50; i++ {
+					// skip ahead again... This should wrap around...
+					w.d.fast.Reset()
+				}
+				w.d.fast.Reset()
+
+				_, err = w.Write(in)
+				if err != nil {
+					t.Fatal(err)
+				}
+				for i := 0; i < (math.MaxUint32-bufferReset)/maxMatchOffset; i++ {
+					// skip ahead to where we are close to wrap around...
+					w.d.fast.Reset()
+				}
+
+				_, err = w.Write(in)
+				if err != nil {
+					t.Fatal(err)
+				}
+				err = w.Close()
+				if err != nil {
+					t.Fatal(err)
+				}
+			}
+		})
 	}
 }
 

--- a/flate/writer_test.go
+++ b/flate/writer_test.go
@@ -279,14 +279,17 @@ func TestWriter_Reset(t *testing.T) {
 	in := buf.Bytes()
 	for l := 0; l < 10; l++ {
 		l := l
+		if testing.Short() && l > 1 {
+			continue
+		}
 		t.Run(fmt.Sprintf("level-%d", l), func(t *testing.T) {
 			t.Parallel()
-			for offset := 1; offset <= 256; offset *= 2 {
+			offset := 1
+			if testing.Short() {
+				offset = 256
+			}
+			for ; offset <= 256; offset *= 2 {
 				// Fail after 'fail' writes
-				if testing.Short() {
-					// Skip...
-					offset = 128
-				}
 				w, err := NewWriter(ioutil.Discard, l)
 				if err != nil {
 					t.Fatalf("NewWriter: level %d: %v", l, err)

